fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -138,100 +137,134 @@ public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull
-                                        Object value = entryKey.get(i + offset + j);
-                                        if (value != null) {
-                                            insertDocumentField(child, entryKey.get(i + offset + j), document, prefix);
-                                        }
-                                    }
-                                }
-                                offset += children.size() - 1;
-                            } else if (expression instanceof LuceneFieldKeyExpression) {
-                                Object value = entryKey.get(i + offset);
-                                if (value != null) {
-                                    insertDocumentField((LuceneFieldKeyExpression)expression, value, document, null);
-                                }
-                            }
-                        }
-                    }
-                } catch (IOException ioe) {
-                    throw new RecordCoreException("Issue updating index keys", "Key", entry, "IndexEntries", indexEntries,
-                            "savedRecord", savedRecord, ioe);
-                }
-            });
-            writer.addDocument(document);
-        } catch (Exception e) {
-            throw new RecordCoreException(e);
-        }
-        return AsyncUtil.DONE;
-    }
 
-    private void insertDocumentField(final LuceneFieldKeyExpression expression, Object value,
-                                     final Document document, String prefix) {
-        if (!(value == null && expression.getType() != LuceneKeyExpression.FieldType.STRING)) {
-            String fieldName = prefix == null ? expression.getFieldName() : prefix.concat("_").concat(expression.getFieldName());
-            switch (expression.getType()) {
-                case INT:
-                    // Todo: figure out how to expand functionality to include storage, sorting etc.
-                    document.add(new IntPoint(fieldName, (Integer)value));
-                    break;
-                case STRING:
-                    document.add(new TextField(fieldName, value == null ? "" : (String)value, expression.isStored() ? Field.Store.YES : Field.Store.NO));
-                    break;
-                case LONG:
-                    document.add(new LongPoint(fieldName, (long) value));
-                    break;
-                default:
-                    throw new RecordCoreArgumentException("Invalid type for lucene index field", "type", expression.getType());
+            // if old record has fields in the index and there are changes then delete old record
+            if (!oldRecordFields.isEmpty() && (!updatedFields.isEmpty() || newRecord == null)) {
+                Query query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(oldRecord.getPrimaryKey().pack()));
+                writer.deleteDocuments(query);
```

Comment:
Is this necessary when we call `addDocument` below? That is, without it, would there be two documents with the same fields or would the one below overwrite?

Reply:
Without it there would be two documents with the same fields. Same as how there can be multiple fields with the exact same key/value.

Reply:
it does not delete the previous one. It passes null to the delNode value in the update method.

Reply:
✅